### PR TITLE
Remove deprecated imsave method in neural style transfer example

### PR DESCRIPTION
**Pull request by stevemurr**
While attempting to run the neural style transfer example, the imsave method failed.  According to https://docs.scipy.org/doc/scipy/reference/generated/scipy.misc.imsave.html#scipy.misc.imsave  
`scipy.misc.imsave` is deprecated and `imageio.imwrite` should be used.  After making this change the script runs successfully.

### CODE REVIEW

**Comment by Dref360** [comment]
This adds a dependency?

**Comment by stevemurr** [comment]
Yes, this adds `imageio` as a dependency.

**Comment by fchollet** [comment]
Does scipy no longer have a built-in way to read/write images?

**Comment by fchollet** [comment]
- Note that the save methods (`imread/imwrite`) are used in several other Keras examples
- In the `keras.preprocessing` module, we rely on PIL to open/save images. In order to avoid new dependencies, maybe we should use that here instead?

**Comment by stevemurr** [comment]
Sounds good.  What do you think about adding the following function to `keras.preprocessing.image`  

```python
def save_img(path, x):
    """Save an image stored as a Numpy Array to a path or file object.

    # Arguments
        path: Path or file object
        x: Input Numpy Array
    """
    img = pil_image.fromarray(x)
    img.save(path)
```

**Comment by fchollet** [comment]
Sounds reasonable. @Dref360 what do you think?

**Comment by Dref360** [comment]
We already have `array_to_img` which would make this feature more UX-friendly.
https://github.com/keras-team/keras/blob/master/keras/preprocessing/image.py#L277

save_img should have a similar API
Use : https://github.com/keras-team/keras/blob/master/keras/preprocessing/image.py#L1446

**Comment by stevemurr** [comment]
Hi,

@Dref360 , I updated the `save_img` method to use `array_to_img` and updated the neural style transfer example to reflect the usage (I pushed an update to this PR that you can check out).  I also tested the change and verified images were saved as expected.  

@fchollet mentioned that numerous examples used the original `imread` and `imwrite` methods so if the `save_img` method looks good I can update and test the other example scripts that made use of the original `scipy.misc.imread` and `scipy.misc.imwrite` methods and update this PR. 

Let me know what you think.

File: examples/neural_style_transfer.py [hunk 1]
```diff
@@ -50,8 +50,7 @@
 '''
 
 from __future__ import print_function
-from keras.preprocessing.image import load_img, img_to_array
-from scipy.misc import imsave
+from keras.preprocessing.image import load_img, save_img, img_to_array
```
Comment (Dref360) [inline examples/neural_style_transfer.py]: img_to_array is not used

File: keras/preprocessing/image.py [hunk 2]
```diff
@@ -358,6 +358,17 @@ def img_to_array(img, data_format=None):
     return x
 
 
+def save_img(path, x):
+    """Save an image stored as a Numpy Array to a path or file object.
```
Comment (Dref360) [inline keras/preprocessing/image.py]: As I said, using img_to_array would be nicer. The API of `save_img` should reflect the one of img_to_array

Comment (stevemurr) [reply keras/preprocessing/image.py]: I forgot to add the change to my last commit - sorry about that.  The `save_img` now reflects the `array_to_img` API.

File: examples/neural_style_transfer.py (hunk 1 above)
Comment (stevemurr) [reply examples/neural_style_transfer.py]: In this script `img_to_array` is used in the `preprocess_image` method here:
https://github.com/keras-team/keras/blob/b09ec1c9bf827622a5eebb94ba0af5bf202ca359/examples/neural_style_transfer.py#L98-L103

File: keras/preprocessing/image.py [hunk 3]
```diff
@@ -358,6 +358,19 @@ def img_to_array(img, data_format=None):
     return x
 
 
+def save_img(path, x, data_format=None, scale=True):
+    """Save an image stored as a Numpy Array to a path or file object.
+
+    # Arguments
+        path: Path or file object.
+        x: Numpy array.
+        data_format: Image data format.
```
Comment (fchollet) [inline keras/preprocessing/image.py]: Explicitly mention that this is 'channels_first'/'channels_last' (since it may be confusing vs. image file format).

File: keras/preprocessing/image.py [hunk 4]
```diff
@@ -358,6 +358,19 @@ def img_to_array(img, data_format=None):
     return x
 
 
+def save_img(path, x, data_format=None, scale=True):
+    """Save an image stored as a Numpy Array to a path or file object.
```
Comment (fchollet) [inline keras/preprocessing/image.py]: Nit: "array" (uncapitalized)

File: keras/preprocessing/image.py [hunk 5]
```diff
@@ -358,6 +358,19 @@ def img_to_array(img, data_format=None):
     return x
 
 
+def save_img(path, x, data_format=None, scale=True):
```
Comment (fchollet) [inline keras/preprocessing/image.py]: I think we should add the arguments "file_format" and "**kwargs" (passed to `Image.save()`) https://pillow.readthedocs.io/en/3.1.x/reference/Image.html#PIL.Image.Image.save

"format" needs to be renamed "file_format" to avoid confusion with "data_format".

File: keras/preprocessing/image.py (hunk 3 above)
Comment (stevemurr) [reply keras/preprocessing/image.py]: I updated the docstring for the `data_format` parameter to explicitly state the options of `{"channels_first" or "channels_last"}`.

File: keras/preprocessing/image.py (hunk 5 above)
Comment (stevemurr) [reply keras/preprocessing/image.py]: Added `file_format` and `**kwargs` parameters which are passed to the `Image.save` method.

**Comment by fchollet** [comment]
PEP8 error: https://travis-ci.org/keras-team/keras/jobs/371795626

**Comment by stevemurr** [comment]
Cool!  I'll check the pep8 error, update the other example scripts and update this PR.

**Comment by stevemurr** [comment]
Hi @fchollet , my PR fails a test regarding the undocumented `**kwargs` parameter passed to `save_img`.  I looked around the codebase and found instances of `**kwargs` that were not documented in their methods docstring.  What is the preferred pattern for documenting `**kwargs`?  I'm thinking I can either 1) Document the `**kwargs` in the docstring (which causes the test to pass) or 2) Disambiguate and document the `**kwargs` parameters passed to `Image.save`.  

Let me know what you think.

File: examples/deep_dream.py [hunk 6]
```diff
@@ -83,7 +84,8 @@ def deprocess_image(x):
 loss = K.variable(0.)
 for layer_name in settings['features']:
     # Add the L2 norm of the features of a layer to the loss.
-    assert layer_name in layer_dict.keys(), 'Layer ' + layer_name + ' not found in model.'
+    assert layer_name in layer_dict.keys(), 'Layer ' + layer_name + \
+        ' not found in model.'
```
Comment (fchollet) [inline examples/deep_dream.py]: Don't include this unrelated change

File: examples/deep_dream.py [hunk 7]
```diff
@@ -67,6 +67,7 @@ def deprocess_image(x):
     x = np.clip(x, 0, 255).astype('uint8')
     return x
 
+
```
Comment (fchollet) [inline examples/deep_dream.py]: Don't include this unrelated change

File: examples/neural_doodle.py [hunk 8]
```diff
@@ -347,6 +348,7 @@ def grads(self, x):
         self.grad_values = None
         return grad_values
 
+
```
Comment (fchollet) [inline examples/neural_doodle.py]: Don't include these unrelated whitespace changes

File: keras/preprocessing/image.py [hunk 9]
```diff
@@ -358,6 +358,28 @@ def img_to_array(img, data_format=None):
...
+    # Arguments
+        path: Path or file object.
+        x: Numpy array.
+        data_format: Image data format.  One of {"channels_first" or "channels_last"}.
+            Default is 'channels_last'.
+        file_format: Optional file format override.  If omitted, the
+            format to use is determined from the filename extension.
+            If a file object was used instead of a filename, this
+            parameter should always be used.
+        scale: Whether to rescale image values to be within [0, 255].
```
Comment (fchollet) [inline keras/preprocessing/image.py]: Please add the line: "**kwargs: Additional keywords arguments passed to `PIL.Image.save()`". This will fix the CI error and it will improve the documentation for users.

**Comment by stevemurr** [comment]
Hi, there seems to be a timeout issue in two of the travis build jobs that's effecting the recent PR's (mine included).  I noticed #10082 is addressing this.  I'll wait until that's resolved then trigger another build.
